### PR TITLE
Fixes GitHub issue #29260 - Metrics instrumenttypes ObservableCounter and ObservableGauge thread safety?

### LINKED ISSUE #29260: Metrics instrumenttypes ObservableCounter and ObservableGauge thread safety?
The instrumenttypes ObservableCounter and ObservableGauge have a callback delegate, that return the values. In the shown examples these values are read without any measure of thread safety, so will the visibility be given? 
Should they be read by `Interlocked.Read` / `Volatile.Read` instead?

For the event counters there's a section [about Concurrency](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/event-counters#concurrency) dealing with that issue. Does this apply for these instrumenttypes too?
If so, a similar section should be added.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 61024acd-c2e8-291a-3007-7f09a1857690
* Version Independent ID: 02ec1bf2-d0fe-dac5-b2c1-45d607430b06
* Content: [Creating Metrics](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/metrics-instrumentation#feedback)
* Content Source: [docs/core/diagnostics/metrics-instrumentation.md](https://github.com/dotnet/docs/blob/main/docs/core/diagnostics/metrics-instrumentation.md)
* Product: **dotnet-fundamentals**
* GitHub Login: @tommcdon
* Microsoft Alias: **tommcdon**

### PR DESCRIPTION
**Pull request by v-thepet**
## Summary

Describe your changes here.

Fixes #29260

Edits article [Creating Metrics](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/metrics-instrumentation#feedback)

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/diagnostics/metrics-instrumentation.md](https://github.com/dotnet/docs/blob/f2207b3705d9c87bc47a0eae4b41e0a9644e99e7/docs/core/diagnostics/metrics-instrumentation.md) | [Creating metrics](https://review.learn.microsoft.com/en-us/dotnet/core/diagnostics/metrics-instrumentation?branch=pr-en-us-37821) |

### CODE REVIEW

File: docs/core/diagnostics/metrics-instrumentation.md [hunk 1]
```diff
@@ -310,37 +310,39 @@ summarize the distribution differently or offer more configuration options.
...
+- Histograms tend to store a lot more data in memory than other metric types. However, the exact memory usage is determined by the collection tool being used.
   If you're defining a large number (>100) of Histogram metrics, you may need to give users guidance not to enable them all at the same time, or to configure their tools to save
   memory by reducing precision. Some collection tools may have hard limits on the number of concurrent Histograms they will monitor to prevent excessive memory use.
 
 - Callbacks for all observable instruments are invoked in sequence, so any callback that takes a long time can delay or prevent all metrics from being collected. Favor
   quickly reading a cached value, returning no measurements, or throwing an exception over performing any potentially long-running or blocking operation.
 
+- The ObservableCounter and ObservableGauge instrument types don't guarantee thread safety. The callbacks occur on a thread that's not usually synchronized with the code that updates the values. When the callback functions are called by multiple threads, it's your responsibility to guarantee the delegates' thread safety.
```
Comment (noahfalk) [inline docs/core/diagnostics/metrics-instrumentation.md:320]:
```suggestion
- The ObservableCounter, ObservableUpDownCounter, and ObservableGauge  callbacks occur on a thread that's not usually synchronized with the code that updates the values. It is your responsibility to either synchronize memory access or accept the inconsistent values that may be observed when using unsynchronized access. Common approaches to synchronize access are using a lock or using <xref:System.Threading.Volatile.Write> and <xref:System.Threading.Volatile.Read>. 
```

What do you think about this wording?

File: docs/core/diagnostics/metrics-instrumentation.md [hunk 2]
```diff
@@ -310,37 +310,39 @@ summarize the distribution differently or offer more configuration options.
...
-    static ObservableCounter<int> s_coatsSold = s_meter.CreateObservableCounter<int>("hatco.store.coats_sold", () => s_rand.Next(1,10));
-    static Random s_rand = new Random();
-
-    static void Main(string[] args)
-    {
-        Console.ReadLine();
-    }
-}
-```
+  ```csharp
```
Comment (noahfalk) [inline docs/core/diagnostics/metrics-instrumentation.md:327]: Is there a reason to indent this? Other sample code above is not similarly indented.

File: docs/core/diagnostics/metrics-instrumentation.md (hunk 1 above)
Comment (gewarren) [reply docs/core/diagnostics/metrics-instrumentation.md:320]: Looks good, but maybe change last sentence to:

"Common approaches to synchronize access are to use a lock or call <xref:System.Threading.Volatile.Write> and <xref:System.Threading.Volatile.Read>. "

File: docs/core/diagnostics/metrics-instrumentation.md (hunk 2 above)
Comment (v-thepet) [reply docs/core/diagnostics/metrics-instrumentation.md:327]: @noahfalk - Yes, it's to align the sample under the bullet. That sample applies only to the preceding bullet point, where it's given as "an example of the problem". If not indented it appears to apply to the whole section, which is briefly confusing if you're skimming.

File: docs/core/diagnostics/metrics-instrumentation.md (hunk 1 above)
Comment (v-thepet) [reply docs/core/diagnostics/metrics-instrumentation.md]: @noahfalk @gewarren - How about another tweak: "It's your responsibility to either synchronize memory access or accept the inconsistent values that can result from using unsynchronized access."

Comment (noahfalk) [reply docs/core/diagnostics/metrics-instrumentation.md]: sounds fine to me!

**Comment by v-thepet** [comment]
#sign-off @gewarren @BillWagner I think this can merge, thanks!

**Review by gewarren — APPROVED** [review]

**Pull request merged**